aten/src/ATen/SparseCsrTensorImpl.cpp

```diff
@@ -0,0 +1,153 @@
+        key_set);
+  }
+}
+} // namespace
+
+SparseCsrTensorImpl::SparseCsrTensorImpl(
+    at::DispatchKeySet key_set,
+    const caffe2::TypeMeta& data_type)
+    : SparseCsrTensorImpl(
+          key_set,
+          data_type,
+          at::empty(
+              {0},
+              at::initialTensorOptions()
+                  .device(SparseCsrTensorSetToDeviceType(key_set))
+                  .dtype(ScalarType::Int)) // crow_indices
+          ,
+          at::empty(
+              {0},
+              at::initialTensorOptions()
+                  .device(SparseCsrTensorSetToDeviceType(key_set))
+                  .dtype(ScalarType::Int)) // col_indices
+          ,
+          at::empty(
+              {0},
+              at::initialTensorOptions()
+                  .device(SparseCsrTensorSetToDeviceType(key_set))
+                  .dtype(data_type)) // values
+      ) {}
+
+SparseCsrTensorImpl::SparseCsrTensorImpl(
+    at::DispatchKeySet key_set,
+    const caffe2::TypeMeta& data_type,
+    at::Tensor crow_indices,
+    at::Tensor col_indices,
+    at::Tensor values)
+    : TensorImpl(key_set, data_type, values.device()),
+      crow_indices_(std::move(crow_indices)),
+      col_indices_(std::move(col_indices)),
+      values_(std::move(values)) {}
+
+void SparseCsrTensorImpl::resize_and_clear_(
+    int64_t nnz_size,
+    IntArrayRef size) {
+  // call crow_indices().options() here since the struct contructor calls the
+  // tensor constructor with args for device specific init.
+  auto empty_crow_indices = at::empty(size[0] + 1, crow_indices().options());
+  auto empty_col_indices = at::empty(nnz_size, col_indices().options());
+  auto empty_values = at::empty(nnz_size, values().options());
+
+  crow_indices_ = empty_crow_indices;
+  col_indices_ = empty_col_indices;
+  values_ = empty_values;
+  sizes_and_strides_.set_sizes(size);
+}
+
+void SparseCsrTensorImpl::resize_as_sparse_csr_tensor_(const Tensor& src) {
+  crow_indices_ = at::empty_like(
+      src.crow_indices(),
+      src.crow_indices().options(),
+      src.crow_indices().suggest_memory_format());
+  col_indices_ = at::empty_like(
+      src.col_indices(),
+      src.col_indices().options(),
+      src.col_indices().suggest_memory_format());
+  values_ = at::empty_like(
+      src.values(),
+      src.values().options(),
+      src.values().suggest_memory_format());
+  sizes_and_strides_.set_sizes(src.sizes());
+}
+
+void SparseCsrTensorImpl::set_member_tensors(
+    const Tensor& crow_indices,
+    const Tensor& col_indices,
+    const Tensor& values) {
+  auto crow_indices_type = crow_indices.scalar_type();
+  auto col_indices_type = col_indices.scalar_type();
+
+  TORCH_CHECK(
+      crow_indices_type == col_indices_type,
+      "both crow_indices and col_indices should have the same type.");
+  TORCH_CHECK(
+      crow_indices_type == kInt || crow_indices_type == kLong,
+      "crow_indices and col_indices must be an int32 or int64 type, but got: ",
```

Comment:
interesting, int32 and int64 indices supported. Should int16 and int8 be supported too? How about unsigned?

Reply:
The support for int32 and int64 originates from the constraints of MKL_INT, but yes, when not using MKL, all integer types could be supported, in principle. The same would apply to COO tensor as well, although it has larger baggage of int64 indices dependencies..

Reply:
What I was expecting to see was that the initial implementation of CSR only supports one index type, and then we try to fry the issue of multiple index support (which is kind of binary size sensitive, since you have to have kernels for the product of index size combinations and element type combinations... maybe? Or I could be wrong haha). But saying we support what MKL supports  fine too, it's just something we have to be careful about